include/boost/math/differentiation/autodiff.hpp

```diff
@@ -659,16 +671,17 @@ template<typename RealType, size_t Order>
 template<typename RealType2, size_t Order2>
 fvar<RealType,Order>& fvar<RealType,Order>::operator*=(const fvar<RealType2,Order2>& cr)
 {
+  using ssize_t = std::make_signed<std::size_t>::type;
   const promote<RealType,RealType2> zero(0);
   if BOOST_AUTODIFF_IF_CONSTEXPR (Order <= Order2)
     for (size_t i=0, j=Order ; i<=Order ; ++i, --j)
-      v[j] = std::inner_product(v.cbegin(), v.cend()-i, cr.v.crbegin()+i, zero);
```

Comment:
Since `std::array<>::difference_type` is signed, this was triggering warnings in MSVC and Clang (with all the all the warnings enabled) due to implicit signed-ness conversions.